libraries/AP_BattMonitor/AP_BattMonitor_DJI.cpp

```diff
@@ -0,0 +1,79 @@
+#include <AP_HAL/AP_HAL.h>
+#include "AP_BattMonitor_DJI.h"
+#include <GCS_MAVLink/GCS.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <AP_Math/crc.h>
+
+/*
+  "battery" monitor for DJI Battery via serial port.
+
+   For future Referance:
+   Total Voltage : 8,9
+   Percentage : 17
+   Cell1 :  20,21
+   Cell2 :  22,23
+   Cell3 :  24,25
+   Cell4 :  26,27
+   
+ */
+extern const AP_HAL::HAL& hal;
+
+ #define DELAY_TIME_US      1000000     // Battery info requesting delay
+
+/// Constructor
+AP_BattMonitor_DJI::AP_BattMonitor_DJI(AP_BattMonitor &mon, AP_BattMonitor::BattMonitor_State &mon_state, AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    const AP_SerialManager &serial_manager = AP::serialmanager();
+    
+    if ((port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_DJIBattery, 0))) {
+        last_send_us = AP_HAL::micros();
+    } else { 
+        gcs().send_text(MAV_SEVERITY_WARNING,"DJIBattery: Port not available or not configured properly");
+    }
+    // need to add check
+    _state.healthy = false;
+}
+
+/*
+  read - read the "voltage" and "percentage"
+*/
+void AP_BattMonitor_DJI::read()
+{
+    if (port != nullptr){
+        uint32_t now = AP_HAL::micros();
+        if (last_send_us != 0 && now - last_send_us > DELAY_TIME_US) {
+           
+            for (uint8_t i=0; i < count; i++) {
+                pktbuf[i] = 0;
+            }
+            
+            uint8_t buf[14] {0xAB, 0x0E, 0x04, 0x22, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xA8}; 
+            port->write(buf, sizeof(buf)); // Requesting Battery info
+            count = 0; // This will reset and sync pktbuf after each reset
+            last_send_us = now;
+        }
+    
+        uint32_t n = port->available();
+    
+        for (uint8_t i=0; i<n; i++) {
+            pktbuf[count++] = port->read();
+            if (count == 37){
```

Comment:
Have you found anything like a start byte? The problem I'm seeing here is if you don't have a start byte to lock onto, losing any bytes on the serial line will cause you to get desynced, and it wouldn't resync until you had lost a multiple of 37 bytes (or worse the CRC matches incorrectly).

A weak solution is if the CRC fails you could throw out the first byte in the buffer and shift everything and try with the next byte, but an 8 bit CRC is weak enough that will be risky for accepting bad data.

Reply:
"count" will reset with after each request and battery will not sent any data without requesting, i think this will only loose a single pack 37 bytes data

Reply:
Lets say you make a request, and the response is supposed to be 37 bytes. If you lose one byte there's 36 bytes in this, and you won't decode it. You then request the next packet as per normal and it sends you 37 bytes as expected. You put the first byte onto the 36 you had before, which gets you to 37, but the CRC then fails. You reset count and read the remaining 36 bytes in, and you are back to the same situation you started with.

Reply:
if first pktbuf will only have 36bytes then it wll not process and the second pktbuf will overwrite the previous, the count will reset just after each request and each request have a 1 sec gap

Reply:
Ah yeah, okay I do see that now. Doesn't help improve the CRC collision case, but you are right it should resync. It seems like you will lose data if the driver takes to long to respond.